### PR TITLE
Custom Force for Version 3

### LINKED ISSUE #866: Port force.constant to the v3 API
## Description

Port force.constant to the v3 API.

## Motivation and context

We should probably retool the interface for force.constant in the process. Also consider removing the unused IMD updater at the same time to remove restrictions on how force.constant needs to behave.

### PR DESCRIPTION
**Pull request by tommy-waltmann**
## Description

This PR converts what was the constant force for v2 into a customizable force in v3. The custom force will allow users to get direct access to the force data buffers in python.

## Motivation and context

We have to port v2 APIs over to v3, but the v2 API was a little to restrictive. Making the force customizable will allow for more flexibility with minimal performance cost.

Resolves #866 

## How has this been tested?

Unit tests are written in `test_custom_force.py`

## Change log

```
Added customizable forces in python.

Changes made in local data context managers now persist after the context manager exits.

```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by tommy-waltmann** [comment]
We've talked about making properties gettable but not settable to ensure users use the `arrays.force[:]` syntax instead of just `arrays.force`. As I look through the code to add that, I see what looks like may have been an attempt to do this already, at line 74 in `local_access.py`. @b-butler Is this what you're intent was when you originally wrote that code?

**Comment by b-butler** [comment]
@tommy-waltmann this was to enable `snap.positions = array` to work by  just setting the data in `snap.positions` to that in `array` exactly like `snap.positions[:] = array`.

**Comment by tommy-waltmann** [comment]
> @tommy-waltmann this was to enable `snap.positions = array` to work by just setting the data in `snap.positions` to that in `array` exactly like `snap.positions[:] = array`.

The way the `__setattr__` is written in `_LocalAccess` does not work as you intend it to. When I attempt to set values in the properties of the local snapshots, those changes are not reflected in cpp once the code exits the context manager. I have explicitly confirmed this with tests.

The local snapshots do not work properly for the same reason we ran into before with the `snap.particles.position` syntax vs the `snap.particles.position[:]` syntax. I looked and there were no tests written for the local snapshots to make sure changes persist after leaving the context manager.

This is a significant bug, and whatever changes are made to fix it should be written so they integrate nicely with both the local snapshot and local force access APIs.

**Comment by tommy-waltmann** [comment]
> > @tommy-waltmann this was to enable `snap.positions = array` to work by just setting the data in `snap.positions` to that in `array` exactly like `snap.positions[:] = array`.
> 
> The way the `__setattr__` is written in `_LocalAccess` does not work as you intend it to. When I attempt to set values in the properties of the local snapshots, those changes are not reflected in cpp once the code exits the context manager. I have explicitly confirmed this with tests.
> 
> The local snapshots do not work properly for the same reason we ran into before with the `snap.particles.position` syntax vs the `snap.particles.position[:]` syntax. I looked and there were no tests written for the local snapshots to make sure changes persist after leaving the context manager.
> 
> This is a significant bug, and whatever changes are made to fix it should be written so they integrate nicely with both the local snapshot and local force access APIs.

Update on this: @b-butler and I had a nice chat and a bug fix is on this branch. We may want to update the tests for the local snapshot API to confirm changes persist after exiting the context manager.

File: hoomd/ForceCompute.cc
Comment (joaander) [inline hoomd/ForceCompute.cc:628]: Document this and any other bugfixes in the proposed change log entry.

**Comment by tommy-waltmann** [comment]
For some reason, there are appear to be tests which involve the const force compute, which we are removing in this PR, in the harmonic bond force ctest file. @joaander Can the old tests be removed in this PR, or is there some reason to keep those around?

File: hoomd/md/force.py
Comment (j-proc) [inline hoomd/md/force.py:221]: Just for my own clarification, although it might be beneficial to users, with the internals of HOOMD, integrators won't touch the GPU so there's no need to set this unless you're doing some other computes with a GPU implementation correct?

File: hoomd/md/force.py
Comment (j-proc) [inline hoomd/md/force.py:257]: This seems like an odd parameter for something that is explicitly set or triggered. Might be helpful to clarify how it's used... which on my own looking up is:
To allow for easy access of the timestep for custom forces which are temporal in nature. Is this a significant enough usecase that we would discourage the perhaps uglier way of the users custom force class keeping a reference to the simulation object and getting it from there?

This is also making me think about efficiency if the force is constant. I'm just wanting to make sure I'm thinking about it correctly. Because the access is O(N(1)) there wouldn't be any benefit from caching these values if they were constant (the simpler old code that we're replacing). It's just a matter of if that those arrays would be kept around internally vs in a python object?

I think in general this is maybe pointing out to me that this might need some documentation about why this isn't based on custom.Action.

File: hoomd/md/pytest/test_custom_force.py
Comment (j-proc) [inline hoomd/md/pytest/test_custom_force.py:39]: Maybe an assert would make sense here? Since we apparently need cupy to run this test anyways and there's not another case I can think of that would trigger it.

File: hoomd/md/pytest/test_custom_force.py
Comment (j-proc) [inline hoomd/md/pytest/test_custom_force.py:67]: If I follow this logic, this is essentially allowing us to later filter out errors which just come from accessing GPU classes on the CPU. It's mainly just for the convenience to write the test once and properly ignore these errors?

File: hoomd/md/pytest/test_custom_force.py
Comment (j-proc) [inline hoomd/md/pytest/test_custom_force.py:298]: I could use a description here because I'm not as familiar with the internals for Ghost particles and MPI. I think it might help other users too when they at correctly accessing these values for MPI runs. I think the main point here is that we don't need to have any portion protected by rank==1, correct?

File: hoomd/data/__init__.py
Comment (j-proc) [inline hoomd/data/__init__.py:9]: Do these belong in the this base level even though force and virial are nonsensical for HPMC. I feel like the answer is probably yes because there's lots of particle data that this is the case for.

File: hoomd/data/local_access_gpu.py
Comment (j-proc) [inline hoomd/data/local_access_gpu.py:8]: What if we build with MD?

File: hoomd/data/__init__.py
Comment (j-proc) [reply hoomd/data/__init__.py:9]: On second thought reading below, I think this could be an issue. I'd say this is the only thing truly blocking approval from me. The other things are clarification

**Review by j-proc — CHANGES_REQUESTED** [review]
@tommy-waltmann, I left quite a few comments that I could use clarification on before I approve and one potential failure point with imports(the only reason I'm requesting changes). 
I did also want to mention though how awesome I think this effort is because you've managed to enable an incredible level of customization and control while also keeping things pretty simple.

File: hoomd/md/force.py
Comment (tommy-waltmann) [reply hoomd/md/force.py:221]: I'm not quite sure what you mean by "integrators won't touch the GPU"; the integration methods have GPU implementations in MD. This property isn't intended to be set. If your simulation uses other computes who need access to the data buffers we expose here, that will still work correctly because the ArrayHandles on the force data buffers are released as soon as the context manager exits.

File: hoomd/md/force.py
Comment (tommy-waltmann) [reply hoomd/md/force.py:257]: The reason the timestep argument is here is because the `compute` method of classes that are derived from the `Compute` class in cpp all take the timestep as an argument, so I just forwarded that argument to the python method.

I don't think there is any reason to be more restrictive than what currently exists in the cpp side of the code. If a user doesn't want to use the timestep argument, they don't have to (in fact almost all computes operate this way).

One of the reasons behind adding the ability to write custom forces in python was so that people could prototype new potentials easily, and since we don't currently have any time-dependent potentials in cpp it seems reasonable to me that someone might want to prototype a time-dependent potential and see what happens.

The reason this isn't just a custom action is because the custom action API doesn't allow a user to change the forces on a set of particles. Using custom actions, you can access the system state via local snapshots, but you can't write your own force compute. That's where custom forces can help.

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:39]: In some of the tests with MPI, there are 0 particles on one of the MPI ranks. If we put an assert here, the test will fail on that rank.

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:67]: This helper function makes sure the simulation fails to run if we are trying to use the GPU custom force class with a CPU simulation device, and otherwise runs the simulation as it is supposed to be run. It is a check that has to be done every time we run a test, so putting all the logic in one helper function is cleaner than writing the same code over and over.

Others can comment if they think there is a cleaner way to write this, I know its somewhat awkward because we have to check the return value after every time we call `_try_running_sim` so the test can exit properly.

File: hoomd/data/local_access_gpu.py
Comment (tommy-waltmann) [reply hoomd/data/local_access_gpu.py:8]: Do you mean what if we build WITHOUT md? That's a good point, and perhaps these class should sit in a `hoomd.md.data` namespace to avoid errors for hpmc only builds.

File: hoomd/data/__init__.py
Comment (tommy-waltmann) [reply hoomd/data/__init__.py:9]: I'll move this to an md specific namespace, see my response to your next comment.

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:298]: The reason for this test is that we have to make sure when we access that `ghost_XXX` and `XXX_with_ghost` attributes of the custom force, they have the right data in them. This test uses an LJ force to make sure there is ghost data available and then just checks to make sure the lengths of the ghost buffers have the right number of particles in them.

I could add a small comment in the docstring for this test to make sure that's clearer. Teaching users how the ghost data is organized and best accessed with custom forces is best left for a tutorial.

File: hoomd/md/pytest/test_custom_force.py
Comment (j-proc) [reply hoomd/md/pytest/test_custom_force.py:67]: I think it might benefit from a more clear name. Especially because `try` is a python keyword. Perhaps something like `run_should_fail` because that's essentially what it returns. But this function is also actually doing the run as well. Will defer to others opinion.

File: hoomd/data/local_access_gpu.py
Comment (j-proc) [reply hoomd/data/local_access_gpu.py:8]: Yes that's what I meant. I couldn't remember if the flag was ENABLE_, BUILD_, etc., thanks for understanding my meaning. Edited for future clarity. 

And @b-butler @joaander are the authorities on the right way to do this. So probably let them weigh in on the correct location for this code to live before making the move.

Comment (joaander) [reply hoomd/data/local_access_gpu.py:8]: HOOMD must import and pass tests when `BUILD_MD=off`. Do not use `try: import ... except: pass` constructions as they mask legitimate errors.

If possible, move code that needs to import `_md` to the `md` component. If not, make the non MD parts of this code work when `_md` is not available. Use the bool `hoomd.version.md_built` to see whether you can import `md` and `_md`.

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:67]: I'll wait until @joaander has a pass over this PR before deciding how to move forward with this comment.

Comment (joaander) [reply hoomd/md/pytest/test_custom_force.py:67]: I agree, this is confusing. It is unclear what code path any given test run is supposed to take and what it is testing.

1) If you are attempting to test that running on the GPU with a CPU device raises an error, write one specific test that does just that.

2) If this is a construction to avoid invalid parameters in the matrix resulting from invalid combinations of two separate parameters, then use a different construction. For example, instead of parameterizing the test on the force class, select the force class based on the device. If that isn't possible, use skips - but skips should be used sparingly. I prefer when they are used only when the user's hardware or software configuration prevents a test from running.

File: hoomd/md/CustomForceCompute.cc
Comment (joaander) [inline hoomd/md/CustomForceCompute.cc:40]:
```suggestion
    if (!m_callback.is(py::none()))
```
This is redundant - make one check or the other.

File: hoomd/md/CustomForceCompute.cc
Comment (joaander) [inline hoomd/md/CustomForceCompute.cc:26]: Set `m_callback` in the constructor.

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [inline hoomd/md/CustomForceCompute.h:59]: These are defined in the base class. Use those instead of duplicating code.

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [inline hoomd/md/CustomForceCompute.h:74]: Where are these used? Why do they need to be duplicated here instead of calling `m_pdata->method()`?

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [inline hoomd/md/CustomForceCompute.h:89]: Local force arrays should be accessible from the base class `Force` object.

File: hoomd/md/data/local_access_cpu.py
Comment (joaander) [inline hoomd/md/data/local_access_cpu.py:27]:
```suggestion
            The reverse tag of a particle. The value
            ``i = force_data.rtag[t]`` is the memory index accessing data
            for the particle with tag ``t``.
```
And similarly for the GPU array class.

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:134]:
```suggestion
    Derive a custom force class from `Custom`, and override the `set_forces` method to  compute forces on particles. Use have direct, zero-copy access to the C++ managed buffers via either the
    `cpu_local_force_arrays` or `gpu_local_force_arrays` property. Choose the
    property that corresponds to the device you wish to alter the data on. In
    addition to zero-copy access to force buffers, custom forces have access to
    the local snapshot API via the ``_state.cpu_local_snapshot`` or the
    ``_state.gpu_local_snapshot`` property. 
    
    See Also:
      See the documentation in `hoomd.State` for more information on the local snapshot API.
```
This will need to be rewrapped and check that the `See Also:` syntax is correct to create a see also info block in the generated docs.

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:137]: Does this need to be said?

File: hoomd/md/force.py
Comment (joaander) [inline hoomd/md/force.py:150]:
```suggestion
                    arrays.force[:] = -5
                    arrays.torque[:] = 3
                    arrays.potential_energy[:] = 27
                    arrays.virial[:] = np.arange(6)[None, :]
```
And similarly for all the other examples.

File: hoomd/md/pytest/test_custom_force.py
Comment (joaander) [inline hoomd/md/pytest/test_custom_force.py:35]: How does this work? I thought we needed the `[:]` syntax to write to the elements of the array?

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! See my comments inline.

File: hoomd/md/CustomForceCompute.h
Comment (tommy-waltmann) [reply hoomd/md/CustomForceCompute.h:59]: Needed to write a version of the getters which are declared `const` and return `const &` in order to make the templating for the `LocalForceComputeData` class to work. I'll try moving the `LocalForceComputeData` to be templated on the `ForceCompute` class before deciding how to address this.

File: hoomd/md/CustomForceCompute.h
Comment (tommy-waltmann) [reply hoomd/md/CustomForceCompute.h:74]: This is more awkwardness to make the templating work. The template class needs to have these methods in order for the `LocalForceComputeData` to compile

**Review by b-butler — COMMENTED** [review]
The local access implementation looks good generally. I have some other comments to address though.

File: hoomd/md/CustomForceCompute.h
Comment (tommy-waltmann) [reply hoomd/md/CustomForceCompute.h:74]: Exposing the rtags could be removed, as Brandon said, but I think the others need to stay so everything can compile.

File: hoomd/md/CustomForceCompute.h
Comment (tommy-waltmann) [reply hoomd/md/CustomForceCompute.h:89]: Why would we want the local force arrays to be read-only? Maybe I'm misunderstanding what you mean, but by "these" you do mean the force arrays, right?

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [reply hoomd/md/CustomForceCompute.h]: Can we just have one method in the base class that is `const`? It should have been that way from the beginning.

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [reply hoomd/md/CustomForceCompute.h]: Read only arrays would prevent users from overwriting forces computed by `pair.LJ` for example.

File: hoomd/md/CustomForceCompute.h
Comment (joaander) [reply hoomd/md/CustomForceCompute.h:4]: Yes, most devs copy and paste existing code. These maintainer lines are a relic from when HOOMD was maintained differently. I have a TODO noted in #1074 to write a pre-commit hook that checks for these and remove them.

File: hoomd/md/pytest/test_custom_force.py
Comment (joaander) [reply hoomd/md/pytest/test_custom_force.py:35]: Should we update the global snapshots to use the same shorthand?

File: hoomd/md/CustomForceCompute.h
Comment (b-butler) [reply hoomd/md/CustomForceCompute.h:4]: Okay

File: hoomd/md/pytest/test_custom_force.py
Comment (b-butler) [reply hoomd/md/pytest/test_custom_force.py:35]: I am fine with either switch. Where setting directly is an error or setting directly is allowed.

File: hoomd/md/CustomForceCompute.h
Comment (tommy-waltmann) [reply hoomd/md/CustomForceCompute.h]: Can do

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:35]: I am a fan of allowing bracketless notation to work, but not advertising it. I can update examples and unit tests to use the bracket notation if need be.

If we enforce the bracket notation, we need to raise errors when users use the bracketless notation so there aren't silent issues with user code.

**Comment by tommy-waltmann** [comment]
@b-butler "You can use test classes for this as I do for the local particle data tests. One class defines the tests and runs on the CPU. The other inherits from the first and runs all test on CPU and GPU."

I only see one class in the file `test_local_snapshots.py` which is used for testing, so your comment confuses me.

File: hoomd/md/pytest/test_custom_force.py
Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:226]: When virials aren't computed because the proper flags aren't set, the `force.virials` property returns `None`. `npt.assert_allclose` can't compare two None values.

Comment (b-butler) [reply hoomd/md/pytest/test_custom_force.py:227]: Why does this not fail then? This should return an array of Booleans.

Comment (tommy-waltmann) [reply hoomd/md/pytest/test_custom_force.py:227]: In this particular test, accessing the virials property just returns `None` because I don't set the flags to compute the virials. I don't know whether that is the desired behavior or not.

**Comment by tommy-waltmann** [comment]
A TODO list for completing this PR, anyone can feel free to add things if they want:

1. Rewrite tests so the test matrix doesn't require so many skips.
2. Write tests for non-custom force classes (like LJ) which ensures getting direct access to their local buffers is read-only.
3. Fix pre-commit issue with importing cupy.
4. Clarify the logic in the `_try_running_sim` test helper function.

**Review by j-proc — APPROVED** [review]
Everything I brought up last time has been addressed or discussed. And it's received quite a bit of attention beyond just getting the functionality in, so aside from making sure the test fails aren't this (it doesn't seem to be) then I think this is a go. 👍 great work @tommy-waltmann

**Review by joaander — APPROVED** [review]
This is great! Thanks!

**Pull request merged**
